File: cmd/chroot.go
package cmd

import (
	"fmt"
	"myproject/utils"

	"github.com/spf13/cobra"
)

var chrootCmd = &cobra.Command{
	Use:   "chroot [mountpoint]",
	Short: "Chroot into a mounted Linux system",
	Args:  cobra.ExactArgs(1),
	Run: func(cmd *cobra.Command, args []string) {
		mountpoint := args[0]
		fmt.Printf("Preparing chroot environment at %s...\n", mountpoint)
		// Mount necessary filesystems
		utils.RunCommandOutput("sudo", "mount", "--bind", "/dev", mountpoint+"/dev")
		utils.RunCommandOutput("sudo", "mount", "--bind", "/proc", mountpoint+"/proc")
		utils.RunCommandOutput("sudo", "mount", "--bind", "/sys", mountpoint+"/sys")
		fmt.Println("Launching chroot shell...")
		out, err := utils.RunCommandOutput("sudo", "chroot", mountpoint)
		if err != nil {
			fmt.Println("Error:", err)
		}
		fmt.Println(out)
	},
}

func init() {
	rootCmd.AddCommand(chrootCmd)
}

File: ghostctl/src/restore/chroot.rs
pub fn enter(mountpoint: &str) {
    use std::process::Command;
    use dialoguer::Confirm;
    println!("Setting up chroot environment at '{}'...", mountpoint);
    if !Confirm::new().with_prompt(format!("Proceed to mount and chroot into '{}'?", mountpoint)).default(false).interact().unwrap() {
        println!("Aborted chroot setup.");
        return;
    }
    // Mount root if not already mounted
    let _ = Command::new("sudo").args(["mount", mountpoint, mountpoint]).status();
    // Bind-mount system dirs
    for dir in ["/dev", "/proc", "/sys", "/run"] {
        let target = format!("{}/{}", mountpoint, dir.trim_start_matches('/'));
        let _ = Command::new("sudo").args(["mount", "--bind", dir, &target]).status();
    }
    // Optionally mount /boot and /efi if present
    for dir in ["/boot", "/efi"] {
        let target = format!("{}/{}", mountpoint, dir.trim_start_matches('/'));
        if std::path::Path::new(&target).exists() {
            let _ = Command::new("sudo").args(["mount", "--bind", dir, &target]).status();
        }
    }
    println!("Launching arch-chroot...");
    let _ = Command::new("sudo").args(["arch-chroot", mountpoint]).status();
    println!("Chroot session ended. Unmounting...");
    // Unmount in reverse order
    for dir in ["/efi", "/boot", "/run", "/sys", "/proc", "/dev"] {
        let target = format!("{}/{}", mountpoint, dir.trim_start_matches('/'));
        let _ = Command::new("sudo").args(["umount", "-l", &target]).status();
    }
    let _ = Command::new("sudo").args(["umount", "-l", mountpoint]).status();
    println!("Chroot environment cleaned up.");
}
